src/BaseKrylov.f90

```diff
@@ -461,4 +462,140 @@ subroutine nonsymmetric_lanczos_tridiagonalization(A, V, W, T, info, kstart, ken
     return
   end subroutine nonsymmetric_lanczos_tridiagonalization
 
+  !-------------------------------------------------------------------------------
+  !-----                                                                     -----
+  !-----     TWO-SIDED ARNOLDI FACTORIZATION FOR GENERAL SQUARE MATRICES     -----
+  !-----                                                                     -----
+  !-------------------------------------------------------------------------------
+
+  subroutine two_sided_arnoldi_factorization(A, V, W, H, G, info, kstart, kend, verbosity, tol)
+    !> Linear operator to be factorized.
+    class(abstract_linop),  intent(in)     :: A
+    !> Left and right Krylov basis.
+    class(abstract_vector), intent(inout)  :: V(:)
+    class(abstract_vector), intent(inout)  :: W(:)
+    !> Upper Hessenberg matrices.
+    real(kind=wp),          intent(inout) :: H(:, :), G(:, :)
+    !> Information flag.
+    integer,                intent(out) :: info
+    !> Optional arguments.
+    integer, optional,       intent(in) :: kstart
+    integer                             :: k_start
+    integer, optional,       intent(in) :: kend
+    integer                             :: k_end
+    logical, optional,       intent(in) :: verbosity
+    logical :: verbose
+    real(kind=wp), optional, intent(in) :: tol
+    real(kind=wp)                       :: tolerance
+    !> Miscellaneous.
+    real(kind=wp)                       :: alpha, beta, gamma, tmp
+    real(kind=wp), allocatable          :: M(:, :), invM(:, :) ! Inner-products matrix and its inverse.
+    real(kind=wp)                       :: tmp_vec(size(V)-1)
+    class(abstract_vector), allocatable :: tmp_krylov_vec
+    integer                             :: i, j, k, kdim
+
+    !> Check Krylov subspace dimensions.
+    if (size(V) .ne. size(W)) then
+       write(*, *) "INFO : Left and right Krylov bases have different dimensions."
+       info = -1
+       return
+    else
+       kdim = size(V) - 1
+    endif
+
+    !> Check the dimensions of the Hessenberg matrices.
+    if (all(shape(H) .ne. [kdim+1, kdim])) then
+       write(*, *) "INFO : Dimensions of the Hessenberg matrix H and Krylov subspace V do not match."
+       info = -2
+       return
+    endif
+
+    if (all(shape(G) .ne. [kdim+1, kdim])) then
+       write(*, *) "INFO : Dimensions of the Hessenberg matrix G and Krylov subspace W do not match."
+       info = -3
+       return
+    endif
+
+    !> Deals with the optional arguments.
+    k_start   = optval(kstart, 1)
+    k_end     = optval(kend, kdim)
+    verbose   = optval(verbosity, .false.)
+    tolerance = optval(tol, rtol)
+
+    !-----------------------------------------------
+    !-----     Two-sided Arnoldi iteration     -----
+    !-----------------------------------------------
+
+    arnoldi : do k = k_start, k_end
```

Comment:
It seems you use the arnoldi routine that can do the full factorization inside a second loop to create the factorisations one step at a time. Is there a reason for this recursion? Why not do the following which seems clearer ?

```
call arnoldi_factorization(A, V, H, info, k_start, k_end, verbose, tolerance, .false.)
call arnoldi_factorization(A, W, G, info, k_start, k_end, verbose, tolerance, .false.)
```
I guess there is a good reason to do it this way. So maybe add a comment explaining it?

Reply:
Sure, `arnoldi_factorization` and all other algs in `BaseKrylov.f90` are implemented the same way.
It is essentially for convenience: the user can either perform a `k`-step Arnoldi factorization in one go, or a step-by-step one if they want compute some intermediate quantities, e.g. checking the number of converged eigenvalues and stop once a desired number has converged.

Basically:
- `call arnoldi_factorization(A, X, H, info)` will compute a full `k`-step factorization (where `k = size(X)-1` is the Krylov subspace dimension).
- Alternatively, you could have
  ```
  do k = 1, size(X)-1
      !> One step of Arnoldi fact.
      call arnoldi_factorization(A, X, H, info, k, k)
      !> Check eigenvalues for instance.
      if ( check(H) .is. .true.) then
          !> Export whatever you want.
          ...
          !> Exit loop.
          break
      endif
  enddo
  ```

Does that make sense? It basically is the same for `two_sided_arnoldi_factorization`. For eigenvalue computations, we want to stop the iteration once the desired number of eigentriplets have converged. For moment-matching reduced-order modeling, you want to do as many iterations as the rank of the low-order model you want to use. The same function can be used in both cases.

Reply:
Yes, that makes sense! Thanks for the explanation!